python/lsst/daf/butler/registry/collections/_base.py

```diff
@@ -0,0 +1,485 @@
+        Name of the column in collections table that identifies it (PK).
+    collectionIdType
+        Type of the PK column in the collections table, one of the
+        `sqlalchemy` types.
+
+    Returns
+    -------
+    spec : `ddl.TableSpec`
+
+    Note
+    ----
+    Assumption here and in the code below is that the name of the identifying
+    column is the same in both collections and run tables. The names of
+    non-identifying columns containing run metadata are fixed.
+    """
+    return ddl.TableSpec(
+        fields=[
+            ddl.FieldSpec(collectionIdName, dtype=collectionIdType, primaryKey=True),
+            TIMESPAN_FIELD_SPECS.begin,
+            TIMESPAN_FIELD_SPECS.end,
+            ddl.FieldSpec("host", dtype=sqlalchemy.String, length=128),
+        ],
+        foreignKeys=[
+            _makeCollForeignKey(collectionIdName, collectionIdName, onDelete="CASCADE"),
+        ],
+    )
+
+
+def makeCollChainTableSpec(collectionIdName: str, collectionIdType: type):
+    """Define specification for "collection_chain" table.
+
+    Parameters
+    ----------
+    collectionIdName : `str`
+        Name of the column in collections table that identifies it (PK).
+    collectionIdType
+        Type of the PK column in the collections table, one of the
+        `sqlalchemy` types.
+
+    Returns
+    -------
+    spec : `ddl.TableSpec`
+
+    Note
+    ----
+    Collection chain is simply an ordered one-to-many relation between
+    collections. The names of the columns in the table are fixed and
+    also hardcoded in the code below.
+    """
+    return ddl.TableSpec(
+        fields=[
+            ddl.FieldSpec("parent", dtype=collectionIdType, primaryKey=True),
+            ddl.FieldSpec("position", dtype=sqlalchemy.SmallInteger, primaryKey=True),
+            ddl.FieldSpec("child", dtype=collectionIdType, nullable=False),
+            ddl.FieldSpec("dataset_type_name", dtype=sqlalchemy.String, length=128, nullable=True),
+        ],
+        foreignKeys=[
+            _makeCollForeignKey("parent", collectionIdName, onDelete="CASCADE"),
+            _makeCollForeignKey("child", collectionIdName),
+        ],
+    )
+
+
+class DefaultCollectionRecord(CollectionRecord):
+    """Default `CollectionRecord` implementation.
+
+    The key (ID) of the record can be the same or different from collection
+    name.
+
+    Parameters
+    ----------
+    id
+        Unique collection ID, can be the same as ``name`` if ``name`` is used
+        for identification.
+    name : `str`
+        Name of the collection.
+    type : `CollectionType`
+        Enumeration value describing the type of the collection.
+    """
+    def __init__(self, id: Any, name: str, type: CollectionType):
+        super().__init__(name, type)
+        self._id = id
+
+    @property
+    def key(self) -> int:
+        # Docstring inherited from CollectionRecord.
+        return self._id
+
+    _id: Any
+    """Database-internal unique collection ID.
+    """
```

Comment:
This may be an indication we should just make `key` a regular attribute (and `__init__` argument) to `CollectionRecord` itself, which would then be concrete and directly usable.

Reply:
Agree, will do.